### PR TITLE
feat(templatemasks): wip proposal

### PR DESCRIPTION
**Pull request by adrienaury**

### CODE REVIEW

**Comment by youen** [comment]
If the same mask with same parameter is used many time in one template all value generated have to be differents.

For the current version  `template: "regex '(a|b|c)'  - regex '(a|b|c)'"` generate only  `"a - a"` ; `"b - b"`; `"c - c"` values

**Comment by adrienaury** [comment]
> If the same mask with same parameter is used many time in one template all value generated have to be differents.
> 
> For the current version `template: "regex '(a|b|c)' - regex '(a|b|c)'"` generate only `"a - a"` ; `"b - b"`; `"c - c"` values

You are right! Fixed in the last commit.

**Comment by adrienaury** [comment]
There is a package dependency loop, so this PR needs to be reworked

File: pkg/templatemask/templatemask.go [hunk 1]
```diff
@@ -32,8 +32,8 @@ type MaskEngine struct {
 }
 
 // NewMask create a MaskEngine
-func NewMask(text string, funcs tmpl.FuncMap) (MaskEngine, error) {
-	temp, err := template.NewEngine(text, funcs)
+func NewMask(text string, funcs tmpl.FuncMap, seed int64) (MaskEngine, error) {
+	temp, err := template.NewEngine(text, funcs, 0)
```
Comment (adrienaury) [inline pkg/templatemask/templatemask.go:36]:
```suggestion
	temp, err := template.NewEngine(text, funcs, seed)
```

File: pkg/regex/regex.go [hunk 2]
```diff
@@ -78,3 +78,15 @@ func Factory(conf model.MaskFactoryConfiguration) (model.MaskEngine, bool, error
 	}
 	return nil, false, nil
 }
+
+func Func(seed int64) interface{} {
+	var callnumber int64
+	return func(regex string) (model.Entry, error) {
+		mask, err := NewMask(regex, seed+callnumber, model.NewSeeder(model.Masking{}, 0))
```
Comment (adrienaury) [inline pkg/regex/regex.go:85]: Fix : build seeder with masking config

File: test/suites/masking_template.yml [hunk 3]
```diff
@@ -53,3 +53,23 @@ testcases:
...
+  - script: |-
+      pimo > result.json <<EOF
+      {"regex": ""}
+      EOF
+    assertions:
+    - result.code ShouldEqual 0
+    - result.systemerr ShouldBeEmpty
+  - script : |-
+      cat > expected.json <<EOF
+      {"regex":"03 97 17 38 57 / 02 35 65 84 76"}
```
Comment (adrienaury) [inline test/suites/masking_template.yml:69]:
```suggestion
      {"regex":"05 30 85 92 14 / 04 40 20 85 81"}
```

File: internal/app/pimo/play.go [hunk 4]
```diff
@@ -135,6 +140,36 @@ func checkSecurityRequirements(pdef model.Definition) error {
 	return nil
 }
 
+func checkMask(mask model.MaskType) error {
```
Comment (youen) [inline internal/app/pimo/play.go:143]: It's the responsibility of the mask to know if it's secure or not. Each time a new mask is added we have to update this function. Maybe we could append a `isSecure()` function on mask interface. and filter on that attribute.

**Review by youen — APPROVED** [review]

File: internal/app/pimo/play.go (hunk 4 above)
Comment (adrienaury) [reply internal/app/pimo/play.go:143]: If we do that, we will filter `template` `add` and `add-transient`  masks, which are very important masks ..

We must know if a mask is secure at execution with parameters (eg: a template mask is insecure only if it contains a call to an insecure mask..)

**Pull request merged**